### PR TITLE
[RFC] Make command spec more declarative

### PR DESCRIPTION
**Pull request by 00vareladavid**
Still a WIP, but I'd be interested if anyone thinks this is a worthwhile change?

My end goal would be a more declarative specification of commands (e.g. specify a list of valid flags for a command , specify the number of expected arguments, ... and have the parser check those declarative constraints).

It will certainly be 'heavier' than the current code but it should be insignificant for interactive input sizes.

### CODE REVIEW

**Comment by KristofferC** [comment]
Yeah, I think this is a good idea and something I have thought about every time I add a new REPL command :).

File: src/REPLMode.jl [hunk 1]
```diff
@@ -58,9 +19,53 @@ end
...
+declare_options = [
+    ("env", :meta, :arg)
+    ,(["project", "p"], :cmd, :switch)
+    ,(["manifest", "m"], :cmd, :switch)
+    ,("major", :cmd, :switch)
+    ,("minor", :cmd, :switch)
+    ,("patch", :cmd, :switch)
+    ,("fixed", :cmd, :switch)
+    ,("coverage", :cmd, :switch)
+    ,("name", :cmd, :switch)
```
Comment (StefanKarpinski) [inline src/REPLMode.jl]: Leading commas like this are anti-idiomatic in Julia; Julia allows a comma after the last element, so just end every line with a comma. Note that vectors can also have no commas at all and just use newlines, but you have to be consistent.

Comment (00vareladavid) [reply src/REPLMode.jl]: Thanks. I'll go with the trailing commas. (Also, I didn't know about the no commas for vectors. Pretty cool!)

Comment (StefanKarpinski) [reply src/REPLMode.jl]: It's technically a different syntax: with commas it's a vector construction syntax, without commas its a vertical concatenation syntax, but in this case the result is the same. If the elements are themselves vectors or matrices, the result can be different, e.g.:
```jl
julia> [[1, 2],
        [3, 4, 5],
        [6, 7]]
3-element Array{Array{Int64,1},1}:
 [1, 2]
 [3, 4, 5]
 [6, 7]

julia> [[1, 2]
        [3, 4, 5]
        [6, 7]]
7-element Array{Int64,1}:
 1
 2
 3
 4
 5
 6
 7
```

Comment (00vareladavid) [reply src/REPLMode.jl]: Ah, I see the difference now: no commas will end up calling `vcat`. Thanks for explaining

**Comment by 00vareladavid** [comment]
There is still more work to be done, but the design is settling a bit more now. I hit some decisions which are best left to you guys.

So far, I have tried to design according to the principle that REPL and the API should be tightly coupled.

Ex: `test --coverage Example` -> `API.test(parse_package("Example"), coverage=true)`
This is the current implementation and I would like it if all REPL commands followed a similar pattern.

### Summary
I would like to offload some of the complexity currently handled by the REPL into the API and introduce a (small) breaking change in the REPL `up` command.

### Details
Ex: `rm --project Example`
AFAICT there is no clean way to translate that to an API call. I would like it `rm --project Example` could be mapped to `API.rm(parse_package("Example"), mode=project)`.

This would simply transfer a small amount of complexity from the REPL to the API. This would have the benefit of being able to reproduce the REPL behavior from the API. Also, I don't think it will introduce breaking changes, but it would add functionality to the API.

Ex: `up --major Example1 --minor`
Admittedly, I don't fully grok `update` yet. But it seems like `up --major Example --minor` would result in different behavior than `up --major Example`.
`up --manifest Example` -> `API.up(parse_package("Example"), mode=PKGMODE_MANIFEST)`
with the Example package having its `mode` field set to `PKGMODE_MANIFEST`
`up --manifest Example --project` -> `API.up(parse_package("Example"), mode=PKGMODE_PROJECT)`
with the Example package having its `mode` field set to `PKGMODE_MANIFEST`

I think the intention was to be able to do: `up --project Example1 --manifest Example2`. I would like it if this would be the command instead: `up --project Example1; up --manifest Example2`. The downside is that a small bit of complexity is offloaded back to the user and the rest to the API. The upside is that `up` will be consistent with the rest of the commands in both parsing and API behavior.

**Comment by 00vareladavid** [comment]
Still some work to be done, but the major parts are in place. I would appreciate some feedback at this point.

### High level view
You can now declare new commands and the parser will take care of enforcing everything. For most cases you only need to supply a thin wrapper around the API and everything should work. My intention is that, for new commands, most of the work is done in the API, then you can just declare the command in `REPLMode.jl`.

The data structure for command declarations is [here](https://github.com/00vareladavid/Pkg.jl/blob/1f9c51511ffea083f82cadd4e2e65cd34d2e996c/src/REPLMode.jl#L871).

A command declaration consists of:
- a list of interchangable names for the command
- a handler function (should be a thin wrapper around the API in most cases)
- an argument specification
    - expected argument types
    - a list of valid number of arguments (an empty list means don't count)
- a list of option declarations

A option declaration consists of:
- a name, or name + short name
- whether an option acts a switch or takes an argument
- a key/value pair

For a key/value pair in an option declaration:
For switch options: k/v pairs and handed directly to the API.
Argument options are a bit more involved: you have the option of applying a function to the option argument and using the result as the value in the k/v pair. Or, if no function is given, just use the raw(`String`) argument as the value.

Meta options are declared [here](https://github.com/00vareladavid/Pkg.jl/blob/1f9c51511ffea083f82cadd4e2e65cd34d2e996c/src/REPLMode.jl#L83). They are structured the same way as command options.

### TODO
- more testing
- light refactoring
- reintroduce support for REPL completions
- reintroduce REPL specific commands (e.g. `help`, `preview`, ...)

**Comment by 00vareladavid** [comment]
Should be ready for a merge. There is a bit more I would like to do (such as not throwing all errors in CommandError), but I would prefer to keep those changes in a separate PR (as this one is already quite big).

cc @StefanKarpinski @KristofferC

**Comment by 00vareladavid** [comment]
Hey guys, I would appreciate any feedback on the status of this PR. The reason I'm asking is that it is becoming painful to keep up to date with the master since `REPLMode.jl` diverges significantly.

As a convincing example in workflow difference you can compare: [my merge](https://github.com/JuliaLang/Pkg.jl/pull/509/commits/93d262967ae4e1e732af2e5caffd1b1145be7f96)
with the [original](https://github.com/JuliaLang/Pkg.jl/commit/66aaa3cbf708a7720bcb56372b4d0155d25a6ad8).

**Comment by KristofferC** [comment]
Sorry, for lack of comments. This looks good but we need to test if it introduces any significant changes to time for first call (e.g. `pkg> st`. 

Likely need to regenerate some of the precompile statements as well but that's on me.

**Comment by 00vareladavid** [comment]
Ok, I think I can test that

**Comment by 00vareladavid** [comment]
I ran some basic tests like this:
```
import Pkg
Pkg.activate("/home/nur0n/pkg/Pkg.jl") # use the same project file
@time Pkg.REPLMode.pkg"st"
```
I think this should test the right thing as the changes for this PR don't take effect until you hit `REPLMode.do_cmd`.

I profiled manually 5 times, then took the average: 
- `master` took about ~5.1 seconds on the first run.
- `refactor/parse` took about ~4.7 seconds on the first run.
- both take ~0.03 seconds for subsequent runs.

Not the most thorough tests, but I think they are reassuring (given that they are testing the right thing...).

**Comment by fredrikekre** [comment]
I haven't looked at the implementation here, but how would "double" commands like `registry add` (https://github.com/JuliaLang/Pkg.jl/issues/453#issuecomment-408879481) fit into this design?

**Comment by StefanKarpinski** [comment]
I wonder if `add` shouldn't just be an implicit shorthand for `package add`...

**Comment by 00vareladavid** [comment]
Excellent point!

The closest thing right now would be to use options:
`registry rm Registry=ae0cb698-197b-42ec-a0a0-4f871aea6013` -> `registry --rm Registry=ae0cb698-197b-42ec-a0a0-4f871aea6013`.
The declaration would be:
```
(       CMD_REGISTRY,
        ["registry"],
        do_registry!,
        (ARG_PKG, []),
        [
            ("add", OPT_SWITCH, :cmd => :add),
            ("rm", OPT_SWITCH, :cmd => :rm),
        ],
        md"",
        )
```

That's the current design though, I hadn't considered explicit support for subcommands. Extending support for them should be easy; it will be my next TODO. What I have in mind is declaring each subcommand:
```
(       CMD_REGISTRY_ADD,
        ["registry_add"],
        do_registry_add!,
        (ARG_PKG, []),
        [],
        md"",
        ),(
        CMD_REGISTRY_RM,
       ["registry_rm"],
       ...
```
And as soon as the parser sees "registry", it will concatenate the next token (e.g. "registry add Example" -> "registry_add Example"). From that point on, the subcommand will behave just like any other command.

**Comment by KristofferC** [comment]
Let's merge this to prevent more merge conflicts and then I can work on fixing up the precompilation if needed.

**Comment by fredrikekre** [comment]
Thanks for adding the `registry` placeholders!

**Comment by 00vareladavid** [comment]
@fredrikekre welcome!

File: src/REPLMode.jl [hunk 2]
```diff
@@ -216,647 +297,364 @@ function parse_quotes(cmd::String)
...
     end
-    API.test(ctx, pkgs; coverage = coverage)
 end
 
-function do_gc!(ctx::Context, tokens::Vector{Token})
-    !isempty(tokens) && cmderror("`gc` does not take any arguments")
-    API.gc(ctx)
+function do_cmd!(command::PkgCommand, repl)
+    meta_opts = APIOptions(command.meta_options, meta_option_specs)
+    ctx = Context(meta_opts...)
```
[the remark this answers is not in the document]
Comment (fredrikekre) [reply src/REPLMode.jl]: Is this still true @KristofferC ? If so, does the help command not work either? It is called after `Context()` both before and after this PR.

[the remark this answers is not in the document]
Comment (00vareladavid) [reply src/REPLMode.jl]: Do you mean when `Base.ACTIVE_PROJECT[] == nothing`? How would I test that failure case? Maybe:
```
Base.ACTIVE_PROJECT[] = nothing
Pkg.REPLMode.pkg"activate"
```

[the remark this answers is not in the document]
Comment (fredrikekre) [reply src/REPLMode.jl]: It is the call to EnvCache that hits https://github.com/JuliaLang/Pkg.jl/blob/e1627a6a5dd7d93aafae545c58d9b584a618ade7/src/Types.jl#L237

Can be reproduced with:

```
julia> using Pkg

julia> empty!(LOAD_PATH)
0-element Array{String,1}

julia> Base.ACTIVE_PROJECT[] = nothing

julia> Base.HOME_PROJECT[] = nothing

julia> Pkg.Types.EnvCache()
ERROR: no active project
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] find_project_file(::Nothing) at /home/fredrik/.julia/dev/Pkg/src/Types.jl:237
 [3] Pkg.Types.EnvCache(::Nothing) at /home/fredrik/.julia/dev/Pkg/src/Types.jl:279 (repeats 2 times)
 [4] top-level scope at none:0

julia> pkg"activate"
ERROR: no active project
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] find_project_file(::Nothing) at /home/fredrik/.julia/dev/Pkg/src/Types.jl:237
 [3] Pkg.Types.EnvCache(::Nothing) at /home/fredrik/.julia/dev/Pkg/src/Types.jl:279 (repeats 2 times)
 [4] Pkg.Types.Context() at ./none:0
 [5] do_cmd!(::Pkg.REPLMode.PkgCommand, ::Pkg.REPLMode.MiniREPL) at /home/fredrik/.julia/dev/Pkg/src/REPLMode.jl:523
 [6] #do_cmd#32(::Bool, ::Function, ::Pkg.REPLMode.MiniREPL, ::String) at /home/fredrik/.julia/dev/Pkg/src/REPLMode.jl:507
 [7] (::getfield(Pkg.REPLMode, Symbol("#kw##do_cmd")))(::NamedTuple{(:do_rethrow,),Tuple{Bool}}, ::typeof(Pkg.REPLMode.do_cmd), ::Pkg.REPLMode.MiniREPL, ::String) at ./none:0
 [8] top-level scope at none:0
```

[the remark this answers is not in the document]
Comment (fredrikekre) [reply src/REPLMode.jl]: And the help command fails too (which it also did before this PR).
```
julia> pkg"help"
ERROR: no active project
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] find_project_file(::Nothing) at /home/fredrik/.julia/dev/Pkg/src/Types.jl:237
 [3] Pkg.Types.EnvCache(::Nothing) at /home/fredrik/.julia/dev/Pkg/src/Types.jl:279 (repeats 2 times)
 [4] Pkg.Types.Context() at ./none:0
 [5] do_cmd!(::Pkg.REPLMode.PkgCommand, ::Pkg.REPLMode.MiniREPL) at /home/fredrik/.julia/dev/Pkg/src/REPLMode.jl:523
 [6] #do_cmd#32(::Bool, ::Function, ::Pkg.REPLMode.MiniREPL, ::String) at /home/fredrik/.julia/dev/Pkg/src/REPLMode.jl:507
 [7] (::getfield(Pkg.REPLMode, Symbol("#kw##do_cmd")))(::NamedTuple{(:do_rethrow,),Tuple{Bool}}, ::typeof(Pkg.REPLMode.do_cmd), ::Pkg.REPLMode.MiniREPL, ::String) at ./none:0
 [8] top-level scope at none:0
```

[the remark this answers is not in the document]
Comment (fredrikekre) [reply src/REPLMode.jl]: So I think we just have to move calls to `do_help!` and `do_activate!` up before creating `Context()`. The help command does not even need the context. `activate` makes use of it, if it exists.

[the remark this answers is not in the document]
Comment (fredrikekre) [reply src/REPLMode.jl]: I'll just include this in #543 while I fix up that PR.

[the remark this answers is not in the document]
Comment (00vareladavid) [reply src/REPLMode.jl]: Just throwing ideas out there. Maybe `ctx.env` could be of type `Union{Nothing, EnvCache}`. Or (I like this more) just get rid of context at the REPL level and instead use a keyword vector. That should work nicely with #515 I think.

[the remark this answers is not in the document]
Comment (fredrikekre) [reply src/REPLMode.jl]: Yea that is probably better. Things like `registry add` won't need the `Context` either.

[the remark this answers is not in the document]
Comment (00vareladavid) [reply src/REPLMode.jl]: Ok, nice catch btw!

[the remark this answers is not in the document]
Comment (fredrikekre) [reply src/REPLMode.jl]: Just to clarify; you mean that the `do_[]!` functions create the `Context` then?

[the remark this answers is not in the document]
Comment (00vareladavid) [reply src/REPLMode.jl]: That would be supported too. I was thinking something like [this](https://github.com/00vareladavid/Pkg.jl/blob/8b9a6c14dcc57ebfa967da583cab142f08e5b098/src/REPLMode.jl#L608). With `Context!` [here](https://github.com/00vareladavid/Pkg.jl/blob/8b9a6c14dcc57ebfa967da583cab142f08e5b098/src/Types.jl#L362). Either way, in the REPL, devs will primarily be dealing with the `Context` (and `APIOptions`) as a dictionary (like [here](https://github.com/00vareladavid/Pkg.jl/blob/8b9a6c14dcc57ebfa967da583cab142f08e5b098/src/REPLMode.jl#L639))

[the remark this answers is not in the document]
Comment (00vareladavid) [reply src/REPLMode.jl]: When we finally hit the API, I think we will end up passing `collect(ctx)` or `Context!(collect(ctx))` depending on the primary way we want users to use the API (I'm not sure which would be better yet).

**Pull request merged**